src/pytorch_lightning/overrides/distributed.py

```diff
@@ -104,6 +105,95 @@ def __iter__(self) -> Iterator[List[int]]:
+                "You seem to have configured a sampler in your DataLoader which"
+                " does not provide `__len__` method. The sampler was about to be"
+                " replaced by `DistributedSamplerWrapper` since `replace_sampler_ddp`"
+                " is True and you are using distributed training. Either provide `__len__`"
+                " method in your sampler, remove it from DataLoader or set `replace_sampler_ddp=False`"
+                " if you want to handle distributed sampling yourself."
+            )
+        if len(sampler) == float("inf"):
+            raise MisconfigurationException(
+                "You seem to have configured a sampler in your DataLoader which"
+                " does not provide finite `__len__` method. The sampler was about to be"
+                " replaced by `DistributedSamplerWrapper` since `replace_sampler_ddp`"
+                " is True and you are using distributed training. Either provide `__len__`"
+                " method in your sampler which returns a finite number, remove it from DataLoader"
+                " or set `replace_sampler_ddp=False` if you want to handle distributed sampling yourself."
+            )
+        self._sampler = sampler
+        self._sampler_list: Optional[List[Any]] = None
+
+    def __getitem__(self, index: int) -> Any:
+        """Gets element of the dataset.
+
+        Args:
+            index: index of the element in the dataset
+        Returns:
+            Single element by index
+        """
+        if self._sampler_list is None:
+            self._sampler_list = list(self._sampler)
+        return self._sampler_list[index]
+
+    def __len__(self) -> int:
+        """
+        Returns:
+            int: length of the dataset
+        """
+        return len(self._sampler)
+
+    def reset(self) -> None:
+        """Reset the sampler list in order to get new sampling."""
+        self._sampler_list = list(self._sampler)
+
+
+class DistributedSamplerWrapper(DistributedSampler):
+    """Wrapper over ``Sampler`` for distributed training. Allows you to use any sampler in distributed mode.
+
+    It is especially useful in conjunction with :class:`torch.nn.parallel.DistributedDataParallel`.
+    In such case, each process can pass a ``DistributedSamplerWrapper`` instance as a ``DataLoader``
+    sampler, and load a subset of subsampled data of the original dataset that is exclusive to it.
+
+    .. note::
+        Sampler is assumed to be of constant size.
+
+    Adapted from https://github.com/catalyst-team/catalyst/blob/v22.04/catalyst/data/sampler.py#L499
+    """
+
+    def __init__(self, sampler: Union[Sampler, Iterable], *args: Any, **kwargs: Any) -> None:
+        super().__init__(_DatasetFromSampler(sampler), *args, **kwargs)
+        self.sampler = sampler
+
+    def __iter__(self) -> Iterator:
+        self.dataset.reset()
+        return iter(self.dataset[index] for index in super().__iter__())
```

Comment:
(Not sure) Could this not re-create the iter? Since you already get a generator which I assume is a freshly created iter

```suggestion
        return (self.dataset[index] for index in super().__iter__())
```

Reply:
You're absolutely right

Reply:
But this way you do create a tuple out of that and the signature says it needs to return an Iterator, so either we actually do need an `iter()` call or we need to adapt the signature...

Reply:
I am fairly certain this is not a tuple but generator expression, which is an iterator by itself https://docs.python.org/3/glossary.html#term-generator-expression